modules/tensor_mechanics/src/actions/GeneralizedPlaneStrainAction.C

```diff
@@ -0,0 +1,123 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "GeneralizedPlaneStrainAction.h"
+
+#include "FEProblem.h"
+#include "Conversion.h"
+
+template<>
+InputParameters validParams<GeneralizedPlaneStrainAction>()
+{
+  InputParameters params = validParams<Action>();
+  params.addClassDescription("Set up the GeneralizedPlaneStrain environment");
+  params.addRequiredParam<std::vector<NonlinearVariableName> >("displacements", "The displacement variables");
+  params.addRequiredParam<std::vector<NonlinearVariableName> >("scalar_strain_zz", "The scalar_strain_zz variables");
+  params.addParam<NonlinearVariableName>("temperature", "The temperature variable");
+  params.addParam<FunctionName>("traction_zz", "0", "Function used to prescribe traction in the out-of-plane Z direction");
+  params.addParam<Real>("factor", 1.0, "Scale factor applied to prescribed traction");
+  params.addParam<bool>("use_displaced_mesh", true, "Whether to use displaced mesh");
+  params.addParam<std::string>("base_name", "Material property base name");
+  params.addParam<std::vector<SubdomainName> >("block", "The list of ids of the blocks (subdomain) that the GeneralizedPlaneStrain kernels will be applied to");
+
+  return params;
+}
+
+GeneralizedPlaneStrainAction::GeneralizedPlaneStrainAction(const InputParameters & params) :
+    Action(params),
+    _displacements(getParam<std::vector<NonlinearVariableName> >("displacements")),
+    _ndisp(_displacements.size()),
+    _block(getParam<std::vector<SubdomainName> >("block")),
+    _scalar_strain_zz(getParam<std::vector<NonlinearVariableName> >("scalar_strain_zz"))
+{
+  if (_ndisp != 2)
+    mooseError("GeneralizedPlaneStrain only works for two dimensional case!");
+
+  for (unsigned int i = 0; i < _scalar_strain_zz.size(); ++i)
+    _scalar_strain_zz_var.push_back(_scalar_strain_zz[i]);
+
+  if (isParamValid("block"))
+    _nblock = _block.size();
+  else
+    _nblock = 1;
+
+  if (_nblock != _scalar_strain_zz.size())
+    mooseError("Number of scalar_strain_zz variable should equal to number of blocks which it will be applied to.");
+}
+
+void
+GeneralizedPlaneStrainAction::act()
+{
+  for (unsigned int i = 0; i < _nblock; ++i)
```

Comment:
I'm not sure why you're creating a separate instance of all of these objects for each block. Can't you just provide the list of blocks to each of the individual objects that you create?